enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,169 @@
+---
+title: upgrade-preflight-checks-for-operators
+authors:
+  - "@LalatenduMohanty"
+reviewers:
+  - "@dgoodwin"
+  - "@crawford"
+  - "@sdodson"
+  - "@wking"
+approvers:
+  - "@crawford"
+  - "@sdodson"
+creation-date: 2020-06-04
+last-updated: 2020-09-02
+status: implementable
+---
+
+# Upgrade Preflight Checks For Operators
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The goal of OpenShift 4 is to make upgrades seamless and risk free.
+To improve the upgrade success we should be able to detect if the new version is compatible with the existing version before running the actual upgrade.
+As OpenShift V4 is composed of many operators, we need to give the operators ability to run some checks before the upgrade to find out if the new incoming version is compatible with the existing version.
+
+## Motivation
+
+CVO managed operators should be able to run preflight checks before going through the actual upgrade.
+It would help operators to find out if they can move to the specific version safely before doing an actual upgrade.
+Also administrators are empowered check if upgrade to the target version is going to be successfully without doing an upgrade.
+Use preflight checks to improve the success rate of upgrades.
+
+
+### Goals
+
+* This enhancement proposes a framework for preflight checks and discusses the CVO side implementation of that framework.
+* Administrators will be able to run preflight checks without running the upgrade.
+* Give Operators (CVO managed) ability to add checks to check if they can safely upgrade to a given release.
```

Comment:
I don't understand the utility of this vs managing this via our existing testing and engineering processes.  We have a graph that describes what you can safely do.  We have a need to break glass and do unsafe things sometimes for customers (force).  Is this trying to make force safer?  Make the graph safer?

Reply:
Sometimes update safety depends on conditions that are visible in-cluster but which are not reported up through Telemetry/Insights.  #426 allows us to address conditions that are reported up via graph tuning, and this proposal covers the bits that aren't.  For example, clusters on restricted networks and those which have opted out of Telemetry/Insights cannot be addressed via #426, and need an in-cluster mechanism like the one in this enhancement.  We could offload checks in those clusters to manual admin operations, but I'd rather have them automated to reduce the chances of folks missing a check and breaking their cluster.

Reply:
> We have a need to break glass and do unsafe things sometimes for customers (force). Is this trying to make force safer? Make the graph safer?

The aim is to make upgrades more predictable and catch things which might be specific to clusters which might prevent successful upgrade and communicating to customers why the upgrade might not be successful or what they need to fix so that upgrade would be successful. Operators now do not have any way to check the environment without actually starting the upgrade. The idea is if we provide a way for operators to check the environment before doing upgrades it would help. AlsoiIt is going to be another layer of safety valve.